File: src/rdruid_analyzer/analysis/talents/tree_of_life.py
from rdruid_analyzer.analysis.attributor import TalentAttributor
from rdruid_analyzer.models.events import HealEvent, ApplyBuffEvent, RemoveBuffEvent
from rdruid_analyzer.tracking.hot_tracker import HotTracker
from rdruid_analyzer.tracking.buff_tracker import BuffTracker

TOL_BUFF = 33891
REJUV_IDS = {774, 155777}
WILD_GROWTH = 48438
IWG_NODE = 82045  # Improved Wild Growth talent node
TICK_WINDOW_MS = 200


class TreeOfLifeAttributor(TalentAttributor):
    """Incarnation: Tree of Life increases healing by 10%, Rejuv by 50%,
    and adds 2 extra Wild Growth targets."""

    name = "Incarnation: Tree of Life"

    def __init__(self):
        super().__init__()
        self._tol_active = False
        self._wg_buffer: list[HealEvent] = []
        self._buffer_start = 0
        self._deferred_wg_healing = 0.0

    def _base_wg_targets(self) -> int:
        return 7 if self.has_talent(IWG_NODE) else 5

    def _flush_wg_buffer(self) -> float:
        if not self._wg_buffer:
            return 0.0
        targets = len({e.target_id for e in self._wg_buffer})
        total_healing = sum(e.amount for e in self._wg_buffer)
        base_targets = self._base_wg_targets()
        # 10% base buff on all ticks
        base_buff = total_healing - total_healing / 1.1
        # Extra target attribution: 2 extra targets on top of base
        if targets > base_targets:
            extra_share = total_healing * (targets - base_targets) / targets
        else:
            extra_share = 0.0
        self._wg_buffer.clear()
        return base_buff + extra_share

    def process_event(self, event, hot_tracker: HotTracker, buff_tracker: BuffTracker):
        if isinstance(event, ApplyBuffEvent) and event.ability_id == TOL_BUFF:
            self._tol_active = True
        elif isinstance(event, RemoveBuffEvent) and event.ability_id == TOL_BUFF:
            self._tol_active = False
            # Flush pending WG buffer when ToL ends
            self._deferred_wg_healing += self._flush_wg_buffer()

    def process_heal(self, event: HealEvent, hot_tracker: HotTracker, buff_tracker: BuffTracker) -> float:
        if not self._tol_active:
            return 0.0

        if event.ability_id in REJUV_IDS:
            return event.amount - event.amount / 1.5

        if event.ability_id == WILD_GROWTH:
            flushed = 0.0
            if self._wg_buffer and event.timestamp - self._buffer_start > TICK_WINDOW_MS:
                flushed = self._flush_wg_buffer()
            if not self._wg_buffer:
                self._buffer_start = event.timestamp
            self._wg_buffer.append(event)
            return flushed

        # All other heals: +10%
        return event.amount - event.amount / 1.1

    def finalize(self) -> float:
        return self._deferred_wg_healing + self._flush_wg_buffer()

File: go-backend/internal/talents/tree_of_life.go
package talents

import (
	"github.com/rdruid-talent-analyzer/go-backend/internal/models"
	"github.com/rdruid-talent-analyzer/go-backend/internal/tracking"
)

const (
	iwgNodeID    = 82045
	tickWindowMS = 200
)

type TreeOfLifeAttributor struct {
	BaseAttributor
	tolActive          bool
	wgBuffer           []*models.HealEvent
	bufferStart        int
	deferredWGHealing  float64
}

func NewTreeOfLifeAttributor() *TreeOfLifeAttributor {
	return &TreeOfLifeAttributor{
		BaseAttributor: NewBaseAttributor("Incarnation: Tree of Life", intPtr(82064), intPtr(103120)),
	}
}

func (a *TreeOfLifeAttributor) baseWGTargets() int {
	if a.HasTalent(iwgNodeID) {
		return 7
	}
	return 5
}

func (a *TreeOfLifeAttributor) flushWGBuffer() float64 {
	if len(a.wgBuffer) == 0 {
		return 0.0
	}
	targets := make(map[int]bool)
	totalHealing := 0
	for _, e := range a.wgBuffer {
		targets[e.TargetID] = true
		totalHealing += e.Amount
	}
	baseTargets := a.baseWGTargets()
	baseBuff := float64(totalHealing) - float64(totalHealing)/TolOtherDivisor
	extraShare := 0.0
	targetCount := len(targets)
	if targetCount > baseTargets {
		extraShare = float64(totalHealing) * float64(targetCount-baseTargets) / float64(targetCount)
	}
	a.wgBuffer = nil
	return baseBuff + extraShare
}

func (a *TreeOfLifeAttributor) ProcessEvent(event models.Event, hot *tracking.HotTracker, buff *tracking.BuffTracker) {
	if ab, ok := event.(*models.ApplyBuffEvent); ok && ab.AbilityID == TreeOfLifeBuff {
		a.tolActive = true
	}
	if rb, ok := event.(*models.RemoveBuffEvent); ok && rb.AbilityID == TreeOfLifeBuff {
		a.tolActive = false
		a.deferredWGHealing += a.flushWGBuffer()
	}
}

func (a *TreeOfLifeAttributor) ProcessHeal(event *models.HealEvent, hot *tracking.HotTracker, buff *tracking.BuffTracker) float64 {
	if !a.tolActive {
		return 0.0
	}

	if event.AbilityID == Rejuvenation || event.AbilityID == GerminationRejuv {
		return float64(event.Amount) - float64(event.Amount)/TolRejuvDivisor
	}

	if event.AbilityID == WildGrowth {
		flushed := 0.0
		if len(a.wgBuffer) > 0 && event.Timestamp-a.bufferStart > tickWindowMS {
			flushed = a.flushWGBuffer()
		}
		if len(a.wgBuffer) == 0 {
			a.bufferStart = event.Timestamp
		}
		a.wgBuffer = append(a.wgBuffer, event)
		return flushed
	}

	return float64(event.Amount) - float64(event.Amount)/TolOtherDivisor
}

func (a *TreeOfLifeAttributor) Finalize() float64 {
	return a.deferredWGHealing + a.flushWGBuffer()
}
